### PR TITLE
provider and date reset

### PR DESCRIPTION
**Pull request by brianseek**
**Note**: Delete the description statements, complete each step. **None are optional**, but can be justified as to why they cannot be completed as written. Provide known gaps to testing that may raise the risk of merging to production.

## Are you removing, renaming or moving a folder in this PR?
- [x] No, I'm not changing any folders (skip to TeamSites and delete the rest of this section)
- [ ] Yes, I'm removing, renaming or moving a folder

If the folder you changed contains a `manifest.json`, search for its `entryName` in the content-build [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json) (the `entryName` there will match).

If an entry for this folder exists in content-build and you are:
1. **Deleting a folder**:
   1. First search `vets-website` for _all_ instances of the `entryName` in your `manifest.json` and remove them in a separate PR. Look particularly for references in `src/applications/static-pages/static-pages-entry.js` and `src/platform/forms/constants.js`. _**If you do not do this, other applications will break!**_
      - _Add the link to your merged vets-website PR here_
   2. Then, Delete the application entry in [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json) and merge that PR **before** this one
      - _Add the link to your merged content-build PR here_

2. **Renaming or moving a folder**: Update the entry in the [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json), but do not merge it until your vets-website changes here are merged. The content-build PR must be merged immediately after your vets-website change is merged in to avoid CI errors with content-build (and Tugboat).

### :warning: TeamSites :warning:
Examples of a TeamSite: https://va.gov/health and https://benefits.va.gov/benefits/. This scenario is also referred to as the "injected" header and footer. You can reach out in the `#sitewide-public-websites` Slack channel for questions.

Did you change site-wide styles, platform utilities or other infrastructure?
- [x] No
- [ ] Yes, and I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#that-sounds-normal-so-whats-the-proxy-all-about) to test the injected header scenario

## Summary

This PR resets both the provider and selected slot in redux and the slot stored in session when someone visits the schedule-referral page. To that end it also updates our mock generators to produce multiple providers and it also swaps the session storage to be slotId based rather than date based. Plus it brings that slot selection into our area of the reducer.

I also made a few changes to get the back button working so that I could manually test this. In the process, I also refactored a few functions into utils to help with readability. 

## Related issue(s)

- _Link to ticket created in va.gov-team repo_
department-of-veterans-affairs/va.gov-team#98701

## Testing done

Manual testing and unit tests.

## What areas of the site does it impact?

Community Care referral scheduling in VAOS

## Acceptance criteria
 - [ ] Provider redux data is retained when moving forward past the date selector page and doesn't refetch when navigating back to page
 - [ ] Slot selection s retained when moving forward past the date selector page and is saved when navigation back to page
 - [ ] When navigating back to the schedule referral page and then forward again, provider is re-fetched and saved slot is reset
 
### Quality Assurance & Testing

- [ ] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ] Linting warnings have been addressed
- [ ] Documentation has been updated ([link to documentation](#) \*if necessary)
- [ ] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/collaboration-cycle/prepare-for-an-accessibility-staging-review) has been performed

### Error Handling

- [ ] Browser console contains no warnings or errors.
- [ ] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Code Review steps
 - turn on `vaOnlineSchedulingCCDirectScheduling` feature locally
 - run mock and app locally
 - navigate to `http://localhost:3001/my-health/appointments/schedule-referral/date-time?id=add2f0f4-a1ea-4dea-a504-a54ab57c6800`
 - select a date/time and then click `continue`
 - click the back link in the bread crumb of the next page
 - observe that the selection is still made and there there was no API call
 - click the back link in the bread crumb again
 - click the `Schedule your appointment` action link
 - observe that the selection is no longer saved and there was an API call to get fresh provider data

### CODE REVIEW

**Review by wullaski — CHANGES_REQUESTED** [review]
Looks good just thought maybe initial values could be null. It would align with the facility initial value.

File: src/applications/vaos/referral-appointments/redux/reducers.js [hunk 1]
```diff
@@ -13,15 +13,18 @@ import {
...
+  SET_INIT_REFERRAL_FLOW,
+  SET_SELECTED_SLOT,
 } from './actions';
 import { FETCH_STATUS } from '../../utils/constants';
 
 const initialState = {
   facility: null,
   sortProviderBy: '',
-  selectedProvider: '',
+  selectedProvider: {},
```
[the remark this answers is not in the document]
Comment (brianseek) [reply src/applications/vaos/referral-appointments/redux/reducers.js:24]: Facility is not something that we need in the reducer anyway. Setting these as their type helps with propType validation in the render cycle.

File: src/applications/vaos/referral-appointments/redux/reducers.js [hunk 2]
```diff
@@ -103,6 +106,18 @@ function ccAppointmentReducer(state = initialState, action) {
...
+      return {
+        ...state,
+        selectedSlot: action.payload,
+      };
+    case SET_INIT_REFERRAL_FLOW:
+      return {
+        ...state,
+        selectedProvider: {},
+        providerFetchStatus: FETCH_STATUS.notStarted,
+        selectedSlot: null,
```
Comment (wullaski) [inline src/applications/vaos/referral-appointments/redux/reducers.js:119]: Should this be empty string to match initial state?

Comment (brianseek) [reply src/applications/vaos/referral-appointments/redux/reducers.js:119]: Yes! I'll update that

Comment (brianseek) [reply src/applications/vaos/referral-appointments/redux/reducers.js]: @wullaski updated

**Comment by brianseek** [comment]
> @ryanshaw / @simiadebowale this seemingly un-related test is failing on this branch but not on main `VAOS Services: Appointment -> isUpcomingAppointmentOrRequest -> should filter future requests:` Any thoughts on what could be causing this?

I debugged and fixed this issue by resetting MockDate and bringing the now var into the describe context.

**Review by juancstlm-a6 — DISMISSED** [review]
Looks good. Small comment on the referral slot key

File: src/applications/vaos/referral-appointments/ScheduleReferral.jsx [hunk 3]
```diff
@@ -5,19 +5,21 @@ import { useDispatch } from 'react-redux';
...
 import ReferralLayout from './components/ReferralLayout';
 import ReferralAppLink from './components/ReferralAppLink';
-import { setFormCurrentPage } from './redux/actions';
+import { setFormCurrentPage, setInitReferralFlow } from './redux/actions';
 
 export default function ScheduleReferral(props) {
   const { currentReferral } = props;
   const location = useLocation();
   const dispatch = useDispatch();
+  const selectedSlotKey = `selected-slot-referral-${currentReferral.UUID}`;
```
[the remark this answers is not in the document]
Comment (brianseek) [reply src/applications/vaos/referral-appointments/ScheduleReferral.jsx:14]: yeah you're right, I will do that

[the remark this answers is not in the document]
Comment (brianseek) [reply src/applications/vaos/referral-appointments/ScheduleReferral.jsx]: @juancstlm-a6 updated!

**Review by juancstlm-a6 — APPROVED** [review]
LGTM

**Comment by brianseek** [comment]
@ryanshaw this is ready for review, here are the review steps:

- turn on vaOnlineSchedulingCCDirectScheduling feature locally
- run mock and app locally
- navigate to http://localhost:3001/my-health/appointments/schedule-referral/date-time?id=add2f0f4-a1ea-4dea-a504-a54ab57c6800
- select a date/time and then click continue
- click the back link in the bread crumb of the next page
- observe that the selection is still made and there there was no API call
- click the back link in the bread crumb again
- click the Schedule your appointment action link
- observe that the selection is no longer saved and there was an API call to get fresh provider data

**Review by wullaski — APPROVED** [review]

**Review by IGallupSoCo — APPROVED** [review]
LGTM

**Pull request merged**